Coprocessor: add fields to the router and the subgraph stages. (#2863)
apollographql/router: .changesets/feat_key_relax_scanner_range.md

### GraphOS Enterprise: Coprocessor read access to request `uri`, `method` and HTTP response status codes ([Issue #2861](https://github.com/apollographql/router/issues/2861), [Issue #2861](https://github.com/apollographql/router/issues/2862))

> We've added the ability for [coprocessors](https://www.apollographql.com/docs/router/customizations/coprocessor) to have read-only access to additional contextual information at [the `RouterService` and SubgraphService`](https://www.apollographql.com/docs/router/customizations/coprocessor/#how-it-works) stages:

Reviewer: Fixing my own typo from the other comment suggestion:

```suggestion
We've added the ability for [coprocessors](https://www.apollographql.com/docs/router/customizations/coprocessor) to have read-only access to additional contextual information at [the `RouterService` and `SubgraphService`](https://www.apollographql.com/docs/router/customizations/coprocessor/#how-it-works) stages:
```

> The `RouterService` stage now has read-only access to the **request** from the client:
>     - `uri` (e.g., `https://host.com/`)
>     - `method` (e.g., `POST`, `GET`)
> 
> The `RouterService` stage now has read-only access to the overall **response** to the client:
>     - `status_code` (e.g. `403`, `200`)
> 
> The `SubgraphService` stage now has read-only access to the **response** of the subgraph request:
>     - `status_code` (e.g., `503`, `200`)

Reviewer: Fixing my own bad (lack of) markdown formatting from the other comment:

```suggestion
The `RouterService` stage now has read-only access to the **request** from the client:
  - `uri` (e.g., `https://host.com/`)
  - `method` (e.g., `POST`, `GET`)

The `RouterService` stage now has read-only access to the overall **response** to the client:
  - `status_code` (e.g. `403`, `200`)

The `SubgraphService` stage now has read-only access to the **response** of the subgraph request:
  - `status_code` (e.g., `503`, `200`)
```

> ### Coprocessor: add fields to the router and the subgraph stage ([Issue #2861](https://github.com/apollographql/router/issues/2861), [Issue #2861](https://github.com/apollographql/router/issues/2862))
> 
> This changeset adds several (read only) fields to coprocessor stages:
> 
> router request:
>     - uri
>     - method
> 
> router response:
>     - status_code
> 
> subgraph response:
>     - status_code
> 
> It removes the`coprocessor.subgraph.all.response.uri` key (that was unused), with a migration.
> 
> It also fixes a bug where a coprocessor operating at the `router_request` stage would fail to deserialize an empty body (which happens for GET requests).

Reviewer: ```suggestion
### GraphOS Enterprise: Coprocessor read access to request `uri`, `method` and HTTP response status codes ([Issue #2861](https://github.com/apollographql/router/issues/2861), [Issue #2861](https://github.com/apollographql/router/issues/2862))

We've added the ability for [coprocessors](https://www.apollographql.com/docs/router/customizations/coprocessor) to have read-only access to additional contextual information at [the `RouterService` and SubgraphService`](https://www.apollographql.com/docs/router/customizations/coprocessor/#how-it-works) stages:

The `RouterService` stage now has read-only access to the **request** from the client:
    - `uri` (e.g., `https://host.com/`)
    - `method` (e.g., `POST`, `GET`)

The `RouterService` stage now has read-only access to the overall **response** to the client:
    - `status_code` (e.g. `403`, `200`)

The `SubgraphService` stage now has read-only access to the **response** of the subgraph request:
    - `status_code` (e.g., `503`, `200`)
```

Reviewer: This is a bit of an omnibus PR that adds functionality, deprecates behavior **AND** fixes a bug. 😄 

I think we should actively avoid that pairing of things in the future, but in the meantime, it's difficult to write a nice changelog for it that really clearly states things for users reading the changelog _(e.g., a Fix as a Feature?, a Deprecation as a Feature?)_.

Can we create additional `.changesets/` files that separate out the changes into their respective sections?

For example?

#### Configuration
```
### Migrate away from unimplemented `coprocessor.subgraph.all.response.uri`

We have removed a completely unimplemented `coprocessor.subgraph.all.response.uri` key from our configuration.  It had no effect, but we will automatically migrate configurations which did use it, resulting in no breaking changes by this removal.

By [@o0ignition0o](https://github.com/o0ignition0o) in https://github.com/apollographql/router/pull/2863
```

### Fixes

```
### Coprocessors: Empty body requests from `GET` requests are now deserialized without error 

Fixes a bug where a coprocessor operating at the `router_request` stage would fail to deserialize an empty body, which is typical for `GET` requests.

By [@o0ignition0o](https://github.com/o0ignition0o) in https://github.com/apollographql/router/pull/2863
```
